src/rapids_singlecell/preprocessing/_neighbors.py

```diff
@@ -356,3 +362,176 @@ def neighbors(
+    metric
+        A known metric's name or a callable that returns a distance.
+    metric_kwds
+        Options for the metric.
+    key_added
+        If not specified, the neighbors data is stored in .uns['neighbors'],
+        distances and connectivities are stored in .obsp['distances'] and
+        .obsp['connectivities'] respectively.
+        If specified, the neighbors data is added to .uns[key_added],
+        distances are stored in .obsp[key_added+'_distances'] and
+        connectivities in .obsp[key_added+'_connectivities'].
+    copy
+        Return a copy instead of writing to adata.
+
+    Returns
+    -------
+    Depending on `copy`, updates or returns `adata` with the following:
+
+    See `key_added` parameter description for the storage path of
+    connectivities and distances.
+
+        **connectivities** : sparse matrix of dtype `float32`.
+            Weighted adjacency matrix of the neighborhood graph of data
+            points. Weights should be interpreted as connectivities.
+        **distances** : sparse matrix of dtype `float32`.
+            Instead of decaying weights, this stores distances for each pair of
+            neighbors.
+    """
+
+    adata = adata.copy() if copy else adata
+    if adata.is_view:
+        adata._init_as_actual(adata.copy())
+    X = _choose_representation(adata, use_rep=use_rep, n_pcs=n_pcs)
+
+    X_contiguous = _check_neighbors_X(X, algorithm)
+    _check_metrics(algorithm, metric)
+
+    n_neighbors = neighbors_within_batch
+
+    if batch_key is None:
+        raise ValueError("Please provide a batch key to perform batch-balanced KNN.")
+
+    if batch_key not in adata.obs:
+        raise ValueError(f"Batch key '{batch_key}' not present in `adata.obs`.")
+
+    n_obs = adata.shape[0]
+    batch_list = adata.obs[batch_key].values
+    batches = np.unique(batch_list)
+    total_neighbors = n_neighbors * len(batches)
+    knn_dist = cp.zeros((X.shape[0], total_neighbors), dtype=np.float32)
+    knn_indices = cp.zeros_like(knn_dist).astype(int)
+
+    for to_ind in range(len(batches)):
+        batch_to = batches[to_ind]
+        mask_to = batch_list == batch_to
+        X_to = X_contiguous[mask_to]
+        ind_to = cp.arange(len(batch_list))[mask_to]
+
+        if algorithm == "brute":
+            sub_ind, sub_dist = _brute_knn(
+                X_to, X_contiguous, n_neighbors, metric, metric_kwds
+            )
+        elif algorithm == "cagra":
+            sub_ind, sub_dist = _cagra_knn(X_to, X_contiguous, n_neighbors, metric)
+        elif algorithm == "ivfflat":
+            sub_ind, sub_dist = _ivf_flat_knn(X_to, X_contiguous, n_neighbors, metric)
+        elif algorithm == "ivfpq":
+            sub_ind, sub_dist = _ivf_pq_knn(X_to, X_contiguous, n_neighbors, metric)
+        col_range = cp.arange(to_ind * n_neighbors, (to_ind + 1) * n_neighbors)
+        knn_indices[:, col_range] = ind_to[sub_ind]
+        knn_dist[:, col_range] = sub_dist
+
```

Comment:
looks like this is a stopping point in the function.

you could wrap the code above into a function returning `knn_indices, knn_dist` and the code below into one returning `distances, connectivities`.

these parts could then be tested individually and would give the code some clearer structure.